packages/worker/src/worker.ts

```diff
@@ -1644,21 +1671,43 @@ export class Worker {
    * @returns the result of `fnOrPromise`
    *
    * Throws on fatal Worker errors.
+   *
+   * **SDK versions `< 1.5.0`**:
+   * This method would not wait for worker to complete shutdown if the inner `fnOrPromise` threw an error.
+   *
+   * **SDK versions `>=1.5.0`**:
+   * This method always waits for both worker shutdown and inner `fnOrPromise` to resolve.
+   * If one of worker run -or- the inner promise throw an error, that error is rethrown.
+   * If both throw an error, a {@link CombinedWorkerRunError} with a `cause` attribute containing both errors.
    */
   async runUntil<R>(fnOrPromise: Promise<R> | (() => Promise<R>)): Promise<R> {
-    const runAndShutdown = async () => {
-      try {
-        if (typeof fnOrPromise === 'function') {
-          return await fnOrPromise();
-        } else {
-          return await fnOrPromise;
+    const workerRunPromise = this.run();
+    const innerPromise = typeof fnOrPromise === 'function' ? fnOrPromise() : fnOrPromise;
+    const [innerResult, workerRunResult] = await Promise.allSettled([
+      innerPromise.finally(() => {
+        if (this.state === 'RUNNING') {
```

Comment:
Wondering if there is a risk of race condition here... What if worker's state is one of `'STOPPING' | 'DRAINING' | 'DRAINED'` at the moment the user's promise throws? Should we wait for the worker to actually reach one of `'STOPPED' | 'FAILED'` state?

Reply:
That's the point of this PR, to make sure the worker is fully stopped before returning from `runUntil`.